src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs

```diff
@@ -22,147 +23,142 @@ internal class ListPackageCommandRunner : IListPackageCommandRunner
     {
         private const string ProjectAssetsFile = "ProjectAssetsFile";
         private const string ProjectName = "MSBuildProjectName";
+        private const int GenericSuccessExitCode = 0;
+        private const int GenericFailureExitCode = 1;
         private Dictionary<PackageSource, SourceRepository> _sourceRepositoryCache;
 
         public ListPackageCommandRunner()
         {
             _sourceRepositoryCache = new Dictionary<PackageSource, SourceRepository>();
         }
 
-        public async Task ExecuteCommandAsync(ListPackageArgs listPackageArgs)
+        public async Task<int> ExecuteCommandAsync(ListPackageArgs listPackageArgs)
         {
+            IReportRenderer reportRenderer = listPackageArgs.Renderer;
+            (int exitCode, ListPackageReportModel reportModel) = await GetReportDataAsync(listPackageArgs);
+            reportRenderer.Render(reportModel);
+            return exitCode;
+        }
+
+        internal async Task<(int, ListPackageReportModel)> GetReportDataAsync(ListPackageArgs listPackageArgs)
+        {
+            // It's important not to print anything to console from below methods and sub method calls, because it'll affect both json/console outputs.
+            var listPackageReportModel = new ListPackageReportModel(listPackageArgs);
             if (!File.Exists(listPackageArgs.Path))
             {
-                Console.Error.WriteLine(string.Format(CultureInfo.CurrentCulture,
+                listPackageArgs.Renderer.AddProblem(problemType: ProblemType.Error,
```

Comment:
An idea (not necessarily for this changeset), would be to somehow not make the logger available in the codepaths that should not have it. 

That'd ensure no weirdness happens.
I'm not sure it's possible, I just don't know how easy of a future change it is :)

Reply:
Sure, but could we address later if such situation rises? I'm not sure what to do this moment, I guess someone is logging error/warning then it should be intentional, also most of the feedbacks I recently addressed was for my future proofing overengineered code removal, don't want to add another one.
Considering all newly added classes are internal and no new public API is introduced it shouldn't be blocker when we refactor next time.

Reply:
This is just a future suggestion. I don't expect it to be done in this PR, so I'll resolve this thread. 

Note that I don't think this feedback and the previous ones are equivalent. 

The whole idea with this rewrite was separating the data being printed from the data being generated, to allow us to easily share code between the json/console implementations. 

Not passing down the logger in the model generation part is part of that idea, no console writing in the model generation.

Reply:
Sure, thank you for your suggestion.